File: operations/src/androidTest/java/l/files/operations/OperationServiceTest.java
package l.files.operations;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;

import java.io.File;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.CountDownLatch;

import l.files.fs.Path;
import l.files.fs.local.LocalPath;
import l.files.operations.OperationService.TaskListener;
import l.files.testing.fs.PathBaseTest;

import static java.lang.System.currentTimeMillis;
import static java.util.Arrays.asList;
import static java.util.Collections.singleton;
import static java.util.concurrent.TimeUnit.SECONDS;
import static l.files.fs.LinkOption.NOFOLLOW;
import static l.files.operations.OperationService.ACTION_CANCEL;
import static l.files.operations.OperationService.EXTRA_TASK_ID;
import static l.files.operations.OperationService.newCancelIntent;
import static l.files.operations.OperationService.newCopyIntent;
import static l.files.operations.OperationService.newDeleteIntent;
import static l.files.operations.OperationService.newMoveIntent;
import static l.files.operations.TaskKind.COPY;
import static l.files.operations.TaskKind.DELETE;
import static l.files.operations.TaskKind.MOVE;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

public final class OperationServiceTest extends PathBaseTest {

    private OperationService service;

    @Override
    protected Path create(File file) {
        return LocalPath.create(file);
    }

    @Override
    protected void setUp() throws Exception {
        super.setUp();
        service = new OperationService();
        service.foreground = false;
    }

    public void test_cancel_intent() throws Exception {

        Intent intent = newCancelIntent(getContext(), 101);
        assertEquals(ACTION_CANCEL, intent.getAction());
        assertEquals(101, intent.getIntExtra(EXTRA_TASK_ID, -1));
        assertEquals(new ComponentName(getContext(), OperationService.class),
                intent.getComponent());
    }

    public void test_cancel_task_not_found() throws Exception {

        TaskListener listener = mock(TaskListener.class);
        service.listener = listener;
        service.onCreate();

        service.onStartCommand(newCancelIntent(getContext(), 1011), 0, 0);

        verify(listener).onNotFound(service, TaskNotFound.create(1011));
    }

    public void test_moves_file() throws Exception {

        Path src = dir1().concat("a").createFile();
        Path dst = dir1().concat("dst").createDir();
        CountDownListener listener = new CountDownListener(MOVE);
        service.listener = listener;
        service.onCreate();

        service.onStartCommand(newMoveIntent(getContext(), singleton(src), dst), 0, 0);

        listener.await();
        assertFalse(src.exists(NOFOLLOW));
        assertTrue(dst.concat(src.name().toPath()).exists(NOFOLLOW));
    }

    public void test_copies_file() throws Exception {

        Path src = dir1().concat("a").createFile();
        Path dst = dir1().concat("dst").createDir();
        CountDownListener listener = new CountDownListener(COPY);
        service.listener = listener;
        service.onCreate();

        service.onStartCommand(newCopyIntent(getContext(), singleton(src), dst), 0, 0);

        listener.await();
        assertTrue(src.exists(NOFOLLOW));
        assertTrue(dst.concat(src.name().toPath()).exists(NOFOLLOW));
    }

    public void test_deletes_files() throws Exception {

        Path a = dir1().concat("a").createFiles();
        Path b = dir1().concat("b/c").createFiles();
        CountDownListener listener = new CountDownListener(DELETE);
        service.listener = listener;
        service.onCreate();

        service.onStartCommand(newDeleteIntent(getContext(), asList(a, b)), 0, 0);

        listener.await();
        assertFalse(a.exists(NOFOLLOW));
        assertFalse(b.exists(NOFOLLOW));
    }

    public void test_task_start_time_is_correct() throws Exception {

        Path file1 = dir1().concat("a").createFile();
        Path file2 = dir1().concat("b").createFile();
        CountDownListener listener = new CountDownListener(DELETE);
        service.listener = listener;
        service.onCreate();

        long start = currentTimeMillis();
        service.onStartCommand(newDeleteIntent(getContext(), asList(file1, file2)), 0, 0);
        listener.await();
        long end = currentTimeMillis();

        Set<Long> times = getTaskStartTimes(listener.getValues());
        assertTrue(times.iterator().next() >= start);
        assertTrue(times.iterator().next() <= end);
    }

    private Set<Long> getTaskStartTimes(List<? extends TaskState> values) {
        Set<Long> result = new HashSet<>();
        for (TaskState value : values) {
            result.add(value.time().time());
        }
        return result;
    }

    private static class CountDownListener implements TaskListener {
        private final CountDownLatch latch;
        private final TaskKind kind;
        private final List<TaskState> values;

        private CountDownListener(TaskKind kind) {
            this(kind, 1);
        }

        private CountDownListener(TaskKind kind, int countDowns) {
            this.latch = new CountDownLatch(countDowns);
            this.values = new ArrayList<>(countDowns);
            this.kind = kind;
        }

        @Override
        public void onUpdate(Context context, TaskState state) {
            values.add(state);
            assertEquals(kind, state.task().kind());
            if (state.isFinished()) {
                latch.countDown();
            }
        }

        public void await() throws InterruptedException {
            assertTrue(latch.await(1, SECONDS));
        }

        public List<TaskState> getValues() {
            return values;
        }

        @Override
        public void onNotFound(Context context, TaskNotFound notFound) {
        }

    }

}

File: operations/src/androidTest/java/l/files/operations/OperationServiceTest.kt
package l.files.operations

import android.content.ComponentName
import android.content.Context
import androidx.test.platform.app.InstrumentationRegistry
import l.files.testing.fs.PathBaseTest
import l.files.testing.fs.Paths.createFiles
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Assert.assertEquals
import org.junit.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import java.nio.file.Files.*
import java.nio.file.LinkOption.NOFOLLOW_LINKS
import java.util.*
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

class OperationServiceTest : PathBaseTest() {

  private lateinit var context: Context
  private lateinit var listener: TaskListener
  private lateinit var service: OperationService

  override fun setUp() {
    super.setUp()
    context = InstrumentationRegistry.getInstrumentation().context
    service = OperationService(false) { listener }
  }

  private fun <T : TaskListener> setListener(listener: T): T {
    this.listener = listener
    return listener
  }

  @Test
  fun cancel_intent() {
    val intent = newCancelIntent(context, 101)
    assertEquals(ACTION_CANCEL, intent.action)
    assertEquals(
      101,
      intent.getIntExtra(EXTRA_TASK_ID, -1).toLong()
    )
    assertEquals(
      ComponentName(context, OperationService::class.java),
      intent.component
    )
  }

  @Test
  fun cancel_task_not_found() {
    listener = mock(TaskListener::class.java)
    service.onCreate()
    service.onStartCommand(newCancelIntent(context, 1011), 0, 0)
    verify(listener).onNotFound(service, TaskNotFound.create(1011))
  }

  @Test
  fun moves_file() {
    val src = createFile(dir1().resolve("a"))
    val dst = createDirectory(dir1().resolve("dst"))
    val listener = setListener(CountDownListener(TaskKind.MOVE))
    service.onCreate()
    service.onStartCommand(newMoveIntent(context, setOf(src), dst), 0, 0)
    listener.await()
    assertThat(exists(src, NOFOLLOW_LINKS), equalTo(false))
    assertThat(
      exists(dst.resolve(src.fileName), NOFOLLOW_LINKS),
      equalTo(true)
    )
  }

  @Test
  fun copies_file() {
    val src = createFile(dir1().resolve("a"))
    val dst = createDirectory(dir1().resolve("dst"))
    val listener = setListener(CountDownListener(TaskKind.COPY))
    service.onCreate()
    service.onStartCommand(newCopyIntent(context, setOf(src), dst), 0, 0)
    listener.await()
    assertThat(exists(src, NOFOLLOW_LINKS), equalTo(true))
    assertThat(
      exists(dst.resolve(src.fileName), NOFOLLOW_LINKS),
      equalTo(true)
    )
  }

  @Test
  fun deletes_files() {
    val a = dir1().resolve("a")
    val b = dir1().resolve("b/c")
    createFiles(a)
    createFiles(b)
    val listener = setListener(CountDownListener(TaskKind.DELETE))
    service.onCreate()
    service.onStartCommand(newDeleteIntent(context, listOf(a, b)), 0, 0)
    listener.await()
    assertThat(exists(a, NOFOLLOW_LINKS), equalTo(false))
    assertThat(exists(b, NOFOLLOW_LINKS), equalTo(false))
  }

  @Test
  fun task_start_time_is_correct() {
    val file1 = createFile(dir1().resolve("a"))
    val file2 = createFile(dir1().resolve("b"))
    val listener = setListener(CountDownListener(TaskKind.DELETE))
    service.onCreate()
    val start = System.currentTimeMillis()
    service.onStartCommand(newDeleteIntent(context, listOf(file1, file2)), 0, 0)
    listener.await()
    val end = System.currentTimeMillis()
    val times = getTaskStartTimes(listener.values)
    assertThat(times.iterator().next() >= start, equalTo(true))
    assertThat(times.iterator().next() <= end, equalTo(true))
  }

  private fun getTaskStartTimes(values: List<TaskState>): Set<Long> =
    values.asSequence().map { it.time().time() }.toSet()

  private class CountDownListener(
    private val kind: TaskKind,
    countDowns: Int = 1
  ) : TaskListener {

    private val latch: CountDownLatch = CountDownLatch(countDowns)

    val values = ArrayList<TaskState>(countDowns)

    override fun onUpdate(context: Context, state: TaskState) {
      values.add(state)
      assertEquals(kind, state.task().kind())
      if (state.isFinished) {
        latch.countDown()
      }
    }

    fun await() {
      assertThat(latch.await(1, TimeUnit.SECONDS), equalTo(true))
    }

    override fun onNotFound(context: Context, notFound: TaskNotFound) {}
  }
}
